src/NuGet.Clients/NuGet.CommandLine/Common/ProtocolDiagnosticsJsonLogger.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.IO;
+using Newtonsoft.Json;
+using NuGet.Protocol.Utility;
+
+namespace NuGet.CommandLine
+{
+    internal sealed class ProtocolDiagnosticsJsonLogger : IDisposable
+    {
+        private object _lock;
+        private StreamWriter _streamWriter;
+        private JsonWriter _jsonWriter;
+
+        internal ProtocolDiagnosticsJsonLogger(string fileName)
+        {
+            _lock = new object();
+            _streamWriter = new StreamWriter(fileName);
+            _jsonWriter = new JsonTextWriter(_streamWriter);
+            _jsonWriter.Formatting = Formatting.Indented;
+            _jsonWriter.WriteStartArray();
+        }
+
+        internal void OnEvent(ProtocolDiagnosticEvent pde)
+        {
+            if (_lock == null)
+            {
+                throw new ObjectDisposedException(nameof(ProtocolDiagnosticsJsonLogger));
+            }
+
+            var timestamp = pde.Timestamp.ToString("o", CultureInfo.InvariantCulture);
+
+            lock (_lock)
+            {
```

Comment:
Curious, why don't you emit even on the start of each request too? I guess you can infer the whole timeline from the duration on the request? Just wondering the rationale.

Reply:
As mentioned in another comment, the goal is to measure/detect slow sources. Raising an event for the beginning of the request, similar to HttpClient's DiagnosticListener, doesn't add value to this goal.

Reply:
Slow can be due to parallelism right? I guess I imagined this would help build a waterfall/gantt chart time model.

Reply:
yes, in my testing, once nuget has to restore more than a few packages, some of the response timings are not believable (I've seen nuget.org's response take more than 4 seconds to process). But this is a separate issue that I've been talking to Damon and other partner teams about. I think as long as the number of packages restored is less than or equal to the number of cpu cores, the timing information is ok/usable. But it's an ongoing effort to understand and try to improve.

Reply:
Okay. If this is shipping in product is there any commitment of not making breaking changes to the event timing or shape?